Feature/consistency and period support

RFreij opened the pull request.

RFreij committed:
Adds PeriodSupport

RFreij committed:
Adds the UndefinedPeriodException

RFreij committed:
Refactor getTotalStats

Use PeriodSupport for both getTotalStats and getDuoStats

Refactor getTotalStats to be consistent with getDuoStats

Removes redundant request from statistics.component

ramon-b commented on api/app/Support/PeriodSupport.php:
```diff
@@ -0,0 +1,130 @@
+            return self::$period();
+        }
+
+        throw (new UndefinedPeriodException($period));
```
Unnecessary parentheses

ramon-b commented on api/app/Support/PeriodSupport.php:
```diff
@@ -0,0 +1,130 @@
+    public static function currentMonth(): stdClass
+    {
+        return (object) [
+            'from' => Carbon::now()->startOfMonth(),
```
A function startOf and a function endOf exists, with the possiblity to pass a string as argument. Therefore, it should be possible to make this dynamic for each period (month, week, etc.).

RFreij replied:
Those functions are added in Carbon 2.0, since laravel 5.5 doesn't support that we can't use them. 
Laravel supports Carbon 2.0 starting from 5.8

![image](https://user-images.githubusercontent.com/15109384/63632882-5d2d4d00-c63f-11e9-99a6-10fd353df992.png)

ramon-b replied:
It seems to be possible to use Carbon 2 with Laravel 5.5 by using https://github.com/kylekatarnls/laravel-carbon-2, but let us refactor it after the API is upgraded to a newer Laravel version.

ramon-b commented on api/app/Http/Controllers/StatsController.php:
```diff
@@ -166,29 +139,22 @@ public function getTotalStats(string $period = '', string $orderBy = 'id', strin
-            }
-        });
+        if ($orderDirection === 'asc') {
+            $data = $data->sortBy($orderBy);
```
Use the second argument of the sortBy function to reduce the code complexity.

RFreij committed:
Fixes comments

RFreij commented:
@ramon-b Fixed comments i was able to fix.

RFreij committed:
Added period class and set return types of PeriodSupport to that period class

RFreij commented:
@ramon-b I've changed it a bit, PeriodSupport now returns a Period instance. This enables us to have autocompletion when using the support class.

I've added it in a seperate commit, so if you don't like it reverting will be easy.

ramon-b commented:
When I try to view the statistics, whereby I order on amount of matches for example (only player id and name works), I get a server error: "asort() expects parameter 2 to be integer, string given" (StatsController:149).

RFreij commented:
> When I try to view the statistics, whereby I order on amount of matches for example (only player id and name works), I get a server error: "asort() expects parameter 2 to be integer, string given" (StatsController:149).

Hmm.. i'll look into that.

RFreij commented:
@ramon-b it seems sortBy's second argument is for a options constant however when I tried this:

```
$options = $orderDirection === 'asc' ? SORT_ASC : SORT_DESC;
$data = $data->sortBy($orderBy, $options);
```

It stil doesn't work. 

This however does work:
```
        $data = $data->sortBy($orderBy, SORT_REGULAR, $orderDirection === 'desc');
```

RFreij committed:
Fixes sorting

ramon-b commented:
Yes, I was thinking of the orderBy function from the query builder. I do not see SORT_ASC and SORT_DESC in the list of sort flags on [https://www.php.net/manual/en/function.sort.php](https://www.php.net/manual/en/function.sort.php), but it is indeed possible to use the third argument of the sortBy function for specifying the direction.

It seems to be working fine now.
